Set up ability to inline CSS into server response.

threehams opened the pull request:
Save a round trip for CSS. Makes a big difference when initial load times are more important than return visits.

threehams committed:
Set up ability to inline CSS into server response.

Save a round trip. Makes a big difference.

sethbattin commented on packages/gluestick/src/renderer/helpers/linkAssets.js:
```diff
@@ -58,24 +75,43 @@ module.exports = function linkAssets(
+        path.join(process.cwd(), 'build', stylesHref),
+      );
+      styleTags.push(
+        <style dangerouslySetInnerHTML={{ __html: contents.toString() }} />,
```
I might be mistaken, but I think recent `react-helmet` allows `style`s with text as children.

threehams replied:
It does, but we're really low-level here (initial server render), and Helmet isn't really designed for order-dependent stuff like styles.

threehams committed:
Declare global instead of suppressing Flow

threehams committed:
Fix CSS inlining when using custom ASSET_URL.

threehams committed:
Use correct publicPath.

threehams committed:
Add additional information in ChunksPlugin.

The name of the chunk can be read in linkAssets, which avoids any
string replacement of public path. Much safer.

threehams committed:
No more [object Object]

threehams committed:
Tests, bug fixes, and code cleanup.

hhtran commented on packages/gluestick/src/config/webpack/ChunksPlugin.js:
```diff
@@ -41,18 +32,18 @@ const getChunksInfoBody = (json: Object, publicPath: string): ChunksInfo => {
   Object.keys(assetsByChunk).forEach((name: string) => {
     // The second asset is usually a source map
-    const jsAsset: string = getAssets(name, 'js')[0];
+    const jsAsset = getAssets(name, 'js')[0];
```
why no more flow typing?

threehams replied:
commented a bit late: most of the Flow types in this file are redundant and already inferred. I improved typing overall by getting rid of some `Object` ones.

threehams commented on packages/gluestick/src/config/webpack/ChunksPlugin.js:
```diff
@@ -22,17 +13,17 @@ const chunkInfoFilePath = (
 
-const getChunksInfoBody = (json: Object, publicPath: string): ChunksInfo => {
-  const assetsByChunk: Object = json.assetsByChunkName;
+const getChunksInfoBody = (stats: Object, publicPath: string): ChunksInfo => {
```
ChunksPlugin is from universal-webpack and was copy/pasted in here a while ago. It tracks JS + CSS chunks and writes the information to a file. I modified it to include the original filename in addition to the URL.

threehams commented on packages/gluestick/src/renderer/helpers/linkAssets.js:
```diff
@@ -1,20 +1,25 @@
 const React = require('react');
 const path = require('path');
 const fs = require('fs');
+// $FlowIgnore promisify is not available in this version of Flow
```
Upgrading Flow will flag this as an error. We can remove it at that time.

threehams commented on packages/gluestick/src/renderer/helpers/linkAssets.js:
```diff
@@ -26,69 +31,85 @@ const filterEntryName = (name: string): string => {
+  const localPath = path.join(process.cwd(), 'build', 'assets', name);
+  const contents = await readFileAsync(localPath);
+
+  cache[name] = contents.toString();
```
This will definitely increase server memory usage, but will not cause garbage collection calls. I'll monitor it after release.

jacobmoretti replied:
Curious how you''ll monitor this. I haven't looked into it too much on NR but do we get pretty good visibility into the health of our node instances through NR?

christinebrass replied:
I think we can leverage datadog for this.

threehams replied:
Today I learned that datadog exists.

threehams commented on packages/gluestick/src/renderer/helpers/linkAssets.js:
```diff
@@ -26,69 +31,85 @@ const filterEntryName = (name: string): string => {
-  let key: number = 0;
-  const entryPointName: string = filterEntryName(entryPoint);
+) {
+  const styleTags = [];
```
I removed a lot of unnecessary Flow typing from this file. All of this is inferred.
